### ISSUE TITLE
Process killed in segmentation pipeline

### ISSUE THREAD

**Issue by shenglandu** [ecker-lab/TreeLearn#6]
Hi,

Thanks for sharing your work. I am following your segmentation pipeline on L1W dataset. However, I got the pipeline process killed during "center chunk" (see the attached figure). I wonder what would possibly be the reason and how I should tune the parameters provided in "pipeline.yaml" to avoid this issue. Thanks a lot in advance for your reply.

Best

![image](https://github.com/ecker-lab/TreeLearn/assets/33870342/6296fc8d-b33b-48e2-91ff-37c0c8b26c78)

**Comment by JanvDelden** [comment]
Hello shenglandu, 

this seems like a surprising error. The 'Killed' signal might be a result of too much RAM usage on your system. While we did not optimize our code to minimize the ram usage, we would not have expected an issue at exactly this position in the script.

Would you mind monitoring htop during running it again and checking the ram usage until the process is killed? 
If ram usage is actually the issue, the easiest solution would be to switch to a system with more ram or make sure that no other processes use ram during running the script. Our code could also be optimized to use less ram, but that would take some work.

**Comment by shenglandu** [comment]
Hi Jan,

Thank you for your detailed reply. I agree that memory usage can be the reason that I got this error. I have followed your tutorials on colab, using plot_7_cut as the pipeline input, and it could work. I also cropped the benchmark L1W into a smaller subset. I could also get the result. So I think the issue is that my machine could not handle the pipeline process of a very large scene. 

I started training TreeLearn with my own dataset. I observed that you are using the pretrained weights from softgroup. I wonder why is this way better than training from scratch (given that pretrained weights are learned from a more general task)? Thanks a lot in advance for your reply, and wish you have a good weekend.


Best,
Shenglan

**Comment by JanvDelden** [comment]
Hi Shenglan,

great to hear you were able to fix the issues. Regarding, the softgroup checkpoint, in our early experiments using it led to better results and faster convergence, so we kept it. If you want to train on your own data, it might also be a good idea to start from one of our checkpoints. We describe our approach and results for that here: https://arxiv.org/pdf/2405.02061
Feel free to ask if there are further questions.

Best regards and have a good weekend as well, 
Jan

**Comment by shenglandu** [comment]
Thanks. I will contact you if I encounter other questions regarding training.

Best Regards,
Shenglan

**Comment by shenglandu** [comment]
Hi Jan,

I hope you are doing well. I am training TreeLearn on a custom dataset, with 8 scenes for training and 1 scene for validation. After 40-50 epochs, the loss becomes Nan and accuracy starts to decrease (as attached below).
![image](https://github.com/ecker-lab/TreeLearn/assets/33870342/d5be6a42-3128-4b88-9115-2fbc64d1dd52)

I picked the checkpoints at epoch 40 since it was the best I could get, with a validation accuracy of 99.75% and offset loss of 1.85. I used this checkpoint in the pipeline, and got pretty good results (as attached).
![image](https://github.com/ecker-lab/TreeLearn/assets/33870342/622aaf3f-258f-4a1b-828c-81216767e7cc)

Therefore, I have three questions regarding TreeLearn:
1. Have you also experienced loss Nan in your training? And what would you think can be the reason? I really have no clue about this.
2. Can you give me suggestions on when to stop training? The epoch number was set to 1400, but my loss starts to be really small at ~40 epoch (0.5 for semantic loss, 1.3 for offset loss). Should I stop here since my loss is really small?
3. Can you also give me suggestions on number of training crops? Currently, I set it to 2k.

Thanks in advance and look forward to your reply.


Best,
Shenglan

**Comment by jonaden94** [comment]
Dear Shenglan,
great to see such good results! Regarding your questions, we recommend the following:

1. We experienced the nan loss on rare occasions. Restarting the training always solved the problem so that we did not bother to find out why this occurs.
2. When fine-tuning from other tree models, we achieved maximum performance when training for 12500 iterations with a batch size of 2 as described [here](https://arxiv.org/pdf/2405.02061). This equals 50 epochs using the default examples_per_epoch=500. In contrast, when training the model from scratch, we found that a large number of epochs is beneficial. I recommend you to monitor the validation loss to see if the offset loss still decreases and keep going as long is it does.
3. Using a larger number of training crops is definitely not harmful for training. So if disc space is not a problem, you can also generate more than 2000. We did not systematically investigate at which point generating more crops does not lead to a better performance anymore.

Best, Jonathan

**Comment by shenglandu** [comment]
Hi Jonathan,

Thanks a lot for your detailed suggestions. 

May I also ask what were the validation semantic loss and offset loss that you obtained for training your model "model_weights_diverse_training_data.pth"?
The reason I am asking is because that, in my training, the best model I could get is with semantic loss of ~0.5 and offset loss of ~1.5. I am not sure if the model has been converged and my losses are small enough.
Thanks in advance for your reply.

Best Regards,
Shenglan

**Comment by jonaden94** [comment]
Dear Shenglan,

I used the L1W dataset as validation for "model_weights_diverse_training_data.pth". The offset loss was around 0.8. I did not record the semantic loss but I would say semantic loss is negligible and usually converges very fast. Since you use another dataset for validation, I do not know if you can expect an equally low offset loss. But intuitively, 1.5 seems a little bit too large to be converged. I suggest you finetune "model_weights_diverse_training_data.pth" for a bit longer (maybe 100-200 epochs with examples_per_epoch=500 and learning rate schedule as described [here](https://arxiv.org/pdf/2405.02061)) and monitor the validation loss to see if it goes lower than when only training for 50 epochs. 

Best, Jonathan

**Comment by shenglandu** [comment]
Hi Jonathan,

Thanks a lot for your message. My offset loss on the training set is 0.95 the lowest. On validation set it's around 1.5~1.8. I use the same training setting as provided in your "train.yaml". If I can solve the loss Nan issue, I will make the network train for more epochs and observe if the network has better convergence.

Best Regards,
Shenglan

**Comment by jonaden94** [comment]
I am confident that you can avoid the nan loss by following the training procedure I recommended to you. If problems remain, feel free to reach out again!

Best, Jonathan

**Comment by shenglandu** [comment]
Thanks! Will reach out to you if I have any problems in future. Wish you have a good summer.

Best Regards,
Shenglan
